<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.335">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Bioinformatics Worskshop 2023 - Day 2</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>

<script src="../site_libs/kePrint-0.0.1/kePrint.js"></script>
<link href="../site_libs/lightable-0.0.1/lightable.css" rel="stylesheet">


</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">Bioinformatics Worskshop 2023</span>
    </a>
  </div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../pages/index.html">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../pages/day1.html">
 <span class="menu-text">Day 1</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="../pages/day2.html" aria-current="page">
 <span class="menu-text">Day 2</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../pages/day3.html">
 <span class="menu-text">Day 3</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/paganilab"><i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com/LabPagani"><i class="bi bi-twitter" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
              <div id="quarto-search" class="" title="Search"></div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#objectives" id="toc-objectives" class="nav-link active" data-scroll-target="#objectives">Objectives</a></li>
  <li><a href="#data-normalization-for-bulk-rna-seq" id="toc-data-normalization-for-bulk-rna-seq" class="nav-link" data-scroll-target="#data-normalization-for-bulk-rna-seq">Data Normalization for bulk RNA-seq</a>
  <ul class="collapse">
  <li><a href="#why-do-we-need-it" id="toc-why-do-we-need-it" class="nav-link" data-scroll-target="#why-do-we-need-it">Why Do We Need It?</a></li>
  <li><a href="#normalization-strategies" id="toc-normalization-strategies" class="nav-link" data-scroll-target="#normalization-strategies">Normalization Strategies</a></li>
  </ul></li>
  <li><a href="#the-deseq2-package" id="toc-the-deseq2-package" class="nav-link" data-scroll-target="#the-deseq2-package">The <code>DESeq2</code> Package</a>
  <ul class="collapse">
  <li><a href="#introduction" id="toc-introduction" class="nav-link" data-scroll-target="#introduction">Introduction</a></li>
  <li><a href="#create-a-deseqdataset-object" id="toc-create-a-deseqdataset-object" class="nav-link" data-scroll-target="#create-a-deseqdataset-object">Create a <code>DESeqDataSet</code> object</a>
  <ul class="collapse">
  <li><a href="#behind-the-design-formula" id="toc-behind-the-design-formula" class="nav-link" data-scroll-target="#behind-the-design-formula">Behind The Design Formula</a></li>
  </ul></li>
  <li><a href="#filtering-genes" id="toc-filtering-genes" class="nav-link" data-scroll-target="#filtering-genes">Filtering genes</a></li>
  <li><a href="#normalizing-count-data" id="toc-normalizing-count-data" class="nav-link" data-scroll-target="#normalizing-count-data">Normalizing Count Data</a></li>
  <li><a href="#transforming-count-data" id="toc-transforming-count-data" class="nav-link" data-scroll-target="#transforming-count-data">Transforming Count Data</a></li>
  <li><a href="#data-quality-visual-assessment" id="toc-data-quality-visual-assessment" class="nav-link" data-scroll-target="#data-quality-visual-assessment">Data Quality Visual Assessment</a>
  <ul class="collapse">
  <li><a href="#clustering" id="toc-clustering" class="nav-link" data-scroll-target="#clustering">Clustering</a></li>
  <li><a href="#sample-to-sample-distances" id="toc-sample-to-sample-distances" class="nav-link" data-scroll-target="#sample-to-sample-distances">Sample-to-sample Distances</a></li>
  <li><a href="#principal-component-analysis-pca" id="toc-principal-component-analysis-pca" class="nav-link" data-scroll-target="#principal-component-analysis-pca">Principal Component Analysis (PCA)</a></li>
  </ul></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Day 2</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<section id="objectives" class="level2">
<h2 class="anchored" data-anchor-id="objectives">Objectives</h2>
<ul>
<li><em>Learn about data normalization</em></li>
<li><em>Learn about the <code>DESeq2</code> package</em></li>
<li><em>Explore different normalization methods</em></li>
<li><em>Normalize the data with functions provided by the <code>DESeq2</code> package</em></li>
<li><em>Perform diagnostic and exploratory analysis on the data</em></li>
</ul>
</section>
<section id="data-normalization-for-bulk-rna-seq" class="level1">
<h1>Data Normalization for bulk RNA-seq</h1>
<section id="why-do-we-need-it" class="level2">
<h2 class="anchored" data-anchor-id="why-do-we-need-it">Why Do We Need It?</h2>
<p>Normalization is the process by which we try to <strong>conform count data to make it comparable across samples and even across different genes</strong>. This is done to ensure that all of the <em>uninteresting</em> differences between genes are minimized in favor of the <strong><em>interesting</em> biological differences</strong> that we want to highlight.</p>
<p>The main factors to take into account when normalizing data are:</p>
<ul>
<li><strong>Sequencing Depth</strong>:</li>
</ul>
<center>
<p><img src="https://hbctraining.github.io/DGE_workshop/img/normalization_methods_depth.png" width="500"></p>
<p><em><strong>Differences in sequencing depth</strong> (total number of reads in a sample) can cause genes to appear differentially expressed just due to technical reasons.</em></p>
</center>
<ul>
<li><strong>Gene Length</strong>:</li>
</ul>
<center>
<p><img src="https://hbctraining.github.io/DGE_workshop/img/normalization_methods_length.png" width="400"></p>
<p><em>Differences in gene length means that <strong>longer genes are prone to have more reads</strong> coming from them, therefore we need to normalize also for this aspect while processing data.</em></p>
</center>
<ul>
<li><strong>RNA Composition</strong>:</li>
</ul>
<center>
<p><img src="https://hbctraining.github.io/DGE_workshop/img/normalization_methods_composition.png" width="500"></p>
<p><em>Differences in RNA composition are also causative of technical artefacts which can skew the analyses. <strong>Differences in the number of genes expressed between samples and the number of differentially expressed gene across samples can greatly and negatively impact normalization methods In the example</strong>, the majority of counts for sample A is related to the DE gene and therefore this might cause other genes to be normalized by this high number of counts, resulting in them appearing to be expressed at a lower level as opposed to the same genes in sample B.</em></p>
</center>
</section>
<section id="normalization-strategies" class="level2">
<h2 class="anchored" data-anchor-id="normalization-strategies">Normalization Strategies</h2>
<p>During the years, many approaches to data normalization have been attempted and are summarized in the table below. You can see that they try to tackle each one of the issues we highlighted above.</p>
<table class="table">
<thead>
<tr class="header">
<th><strong>Normalization Method</strong></th>
<th><strong>Accounted Factors</strong></th>
<th><strong>Description</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>CPM (counts per million)</td>
<td>Sequencing depth</td>
<td>Counts scaled by total read number</td>
</tr>
<tr class="even">
<td>TPM (transcripts per million)</td>
<td>Sequencing depth and gene length</td>
<td>Counts per length of transcript (kb) per million mapped reads</td>
</tr>
<tr class="odd">
<td>FPKM/RPKM</td>
<td>Sequencing depth and gene length</td>
<td>Counts per kilobase of exon mapped reads per million mapped reads</td>
</tr>
<tr class="even">
<td><em>DESeq2</em>’s median of ratios<sup><a href="https://bioconductor.org/packages/release/bioc/html/DESeq2.html">1</a></sup></td>
<td>Sequencing depth and RNA composition</td>
<td>Counts are divided by a sample-specific size factor</td>
</tr>
<tr class="odd">
<td><em>edgeR</em>’s trimmed mean of M values<sup><a href="https://bioconductor.org/packages/release/bioc/html/edgeR.html">2</a></sup></td>
<td>Sequencing depth, RNA composition and gene length</td>
<td>Weighted trimmed mean of the log ratio of expression between samples</td>
</tr>
</tbody>
</table>
<p>Out of all these, we will use one of the more advanced ones provided in the <code>DESeq2</code> package which will be now introduced.</p>
</section>
</section>
<section id="the-deseq2-package" class="level1">
<h1>The <code>DESeq2</code> Package</h1>
<section id="introduction" class="level2">
<h2 class="anchored" data-anchor-id="introduction">Introduction</h2>
<p>One of the main interests behind performing a bulk RNA experiment is <strong>understanding which genes are more or less expressed across a set of conditions of interest</strong>, so we will <strong>compare</strong> gene expression levels and <strong>statistically assess</strong> and quantify differences arising between the conditions represented by our categories of samples. In this section we will start getting a feel for what DESeq2 is and how to use it to perform normalization and differential expression analysis on our bulk RNA seq data.</p>
<blockquote class="blockquote">
<p>💡 <strong>Detailed explanations of the statistical procedures implemented in the package are available in the package’s <a href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html"><em>vignette</em></a>.</strong></p>
</blockquote>
<div class="cell" data-layout-align="center">

</div>
<p>We will start by <strong>loading the package</strong> that we installed in yesterday’s initial “setup” section:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the package</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(DESeq2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The package <strong>contains many <em>functions</em></strong> which are very helpful when dealing with tables of <em>count data</em>, we can inspect them all by typing the following command:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># This should open a popup window in the lower right part of the screen displaying the functions in the package</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>??DESeq</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In the sections below, we will follow the standard analysis workflow suggested by the DESeq2 developers.</p>
</section>
<section id="create-a-deseqdataset-object" class="level2">
<h2 class="anchored" data-anchor-id="create-a-deseqdataset-object">Create a <code>DESeqDataSet</code> object</h2>
<p>In order for the package to <em>read</em> and <em>understand</em> our data and correctly perform the analysis, we need to <strong>organize our data in a way that the functions of the package can handle</strong>. This new object that we are going to create is called <code>DESeqDataSet</code> and there is a utility function to create one starting from the <strong>ingredients</strong> we currently have, (1) a <strong>table of counts</strong> (our <code>matrix</code>), (2) a <strong>table with sample information</strong> (our <code>coldata</code>) and (3) one last thing that we need to decide in order tell the package what comparisons we value the most, this is called a <strong>design formula</strong>.</p>
<section id="behind-the-design-formula" class="level3">
<h3 class="anchored" data-anchor-id="behind-the-design-formula">Behind The Design Formula</h3>
<p>The design formula should <strong>contain the name of a column of interest in our table of samples</strong> which stores the information related to the <strong><em>levels</em></strong> (or <em>categories</em>) we want to contrast. Let’s say that we have a dataset with two conditions (<code>condition_1</code> vs <code>condition_2</code>) that we want to compare. The samples table will look like this, with three replicates for each of the two conditions:</p>
<table class="table">
<thead>
<tr class="header">
<th>Sample Code</th>
<th>Patient</th>
<th>Condition</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>SA1</td>
<td>Pt1</td>
<td>Condition_1</td>
</tr>
<tr class="even">
<td>SA2</td>
<td>Pt2</td>
<td>Condition_1</td>
</tr>
<tr class="odd">
<td>SA3</td>
<td>Pt3</td>
<td>Condition_1</td>
</tr>
<tr class="even">
<td>SA4</td>
<td>Pt4</td>
<td>Condition_2</td>
</tr>
<tr class="odd">
<td>SA5</td>
<td>Pt5</td>
<td>Condition_2</td>
</tr>
<tr class="even">
<td>SA6</td>
<td>Pt6</td>
<td>Condition_2</td>
</tr>
</tbody>
</table>
<p>If we are interested in performing a differential expression analysis comparing <code>condition_1</code> versus <code>condition_2</code>, then our design formula should specify the <code>Condition</code> column.</p>
<blockquote class="blockquote">
<p>💡 <strong>What is the column that we are interested in when specifying the design formula using in our <code>samples</code> table?</strong></p>
</blockquote>
<p>Now that we also understand the design formula, we can create the <code>DESeqDataSet</code> object with the data that we loaded beforehand, but first we need to check that the columns of the <code>counts</code> table are in the same order of the rows of the <code>sample</code> table, this is important since we want to be sure that the right levels of expression are associated to the right sample.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">all</span>(<span class="fu">rownames</span>(samples) <span class="sc">==</span> <span class="fu">colnames</span>(counts))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] TRUE</code></pre>
</div>
</div>
<p>Now we are sure about that we can actually build the object:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a variable containing the design formula</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>design_formula <span class="ot">&lt;-</span>  <span class="er">~</span> Donor <span class="sc">+</span> SampleGroup </span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a `DESeqDataSet` object and call it dds</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span>(<span class="at">countData =</span> counts, </span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>                              <span class="at">colData =</span> samples, </span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>                              <span class="at">design =</span> design_formula)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Great! You have created a <code>DESeqDataSet</code> object which we called <code>dds</code>, this contains all the information related to the counts table and the sample information table in one spot.</p>
</section>
</section>
<section id="filtering-genes" class="level2">
<h2 class="anchored" data-anchor-id="filtering-genes">Filtering genes</h2>
<p>We can the inspect the size of our <code>dds</code> object to check how many genes and samples we have, you can see that this mirrors exactly the <code>counts</code> table that we had previously before incorporating it into the object.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(<span class="fu">counts</span>(dds))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 62854    16</code></pre>
</div>
</div>
<blockquote class="blockquote">
<p>💡 <strong>Notice how we used <code>DESeq2</code>’s function <code>counts()</code> to access the table of gene counts within the <code>DESeqDataSet</code> object.</strong></p>
</blockquote>
<p>Now we can <strong>start removing some genes which are expressed at low levels</strong>. This action is justified both from the biological and the statistical points of view. Biologically, lowly expressed genes are less interesting while statistically, they do not provide enough sampling power to confidently test them for differential expression. In the following chunk we filter our <code>dds</code> object by <strong>keeping only genes (rows) which have at least 10 detected counts when summing across all samples</strong>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> dds[<span class="fu">rowSums</span>(<span class="fu">counts</span>(dds)) <span class="sc">&gt;=</span> <span class="dv">10</span>,]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<blockquote class="blockquote">
<p>💡 <strong>This is somewhat of an “easy” filtering strategy, there are many more complex ones available but sometimes it’s best to keep things simple!</strong></p>
</blockquote>
<p>Let’s check how many genes we’re left with after filtering:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(dds)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 31737    16</code></pre>
</div>
</div>
<p>You can see that <strong>we have cut the number of genes in our dataset by almost half</strong>, now we have 31737 genes.</p>
</section>
<section id="normalizing-count-data" class="level2">
<h2 class="anchored" data-anchor-id="normalizing-count-data">Normalizing Count Data</h2>
<p>As we have discussed above, <strong>normalization is an integral step to the downstream analysis of bulk RNA-seq data</strong>. In this section we will normalize our data using the <code>estimateSizeFactors</code> function of the package. This function creates a value for every gene (a <em>pseudovalue</em>) by taking the geometric mean of expression values across samples for that specific gene and then dividing each value by this new computed <em>pseudovalue</em>. This function returns a <strong>normalized table of values</strong> which is identical in shape to our original <code>counts</code> table as we will see.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Call the function to normalize count data</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">estimateSizeFactors</span>(dds)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can check the values of the computed <strong>size factors</strong> by doing the following, note how <strong>there are as many size factors as there are samples</strong>:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sizeFactors</span>(dds)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">

<table class="table" style="margin-left: auto; margin-right: auto;">
 <thead>
  <tr>
   <th style="text-align:left;">   </th>
   <th style="text-align:right;"> . </th>
  </tr>
 </thead>
<tbody>
  <tr>
   <td style="text-align:left;"> BSSE_QGF_204446 </td>
   <td style="text-align:right;"> 0.9679699 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> BSSE_QGF_204447 </td>
   <td style="text-align:right;"> 0.8817713 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> BSSE_QGF_204448 </td>
   <td style="text-align:right;"> 1.2034053 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> BSSE_QGF_204449 </td>
   <td style="text-align:right;"> 1.0502422 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> BSSE_QGF_204450 </td>
   <td style="text-align:right;"> 1.2141002 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> BSSE_QGF_204451 </td>
   <td style="text-align:right;"> 1.0913873 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> BSSE_QGF_204452 </td>
   <td style="text-align:right;"> 1.1474634 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> BSSE_QGF_204453 </td>
   <td style="text-align:right;"> 1.0950554 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> BSSE_QGF_204458 </td>
   <td style="text-align:right;"> 0.9270823 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> BSSE_QGF_204459 </td>
   <td style="text-align:right;"> 0.8959534 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> BSSE_QGF_204460 </td>
   <td style="text-align:right;"> 0.8036038 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> BSSE_QGF_204461 </td>
   <td style="text-align:right;"> 0.8165707 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> BSSE_QGF_204462 </td>
   <td style="text-align:right;"> 0.9341968 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> BSSE_QGF_204463 </td>
   <td style="text-align:right;"> 1.0681170 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> BSSE_QGF_204464 </td>
   <td style="text-align:right;"> 1.1779602 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> BSSE_QGF_204465 </td>
   <td style="text-align:right;"> 1.0230975 </td>
  </tr>
</tbody>
</table>

</div>
</div>
<p>The results of the normalization are accessible by using the following code, we will save the normalized values in a different variable that we will call <code>normalized_counts</code>, <code>normalized = TRUE</code> tells <code>DESeq2</code> to fetch the normalized version of the count table that we have just created above:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>normalized_counts <span class="ot">&lt;-</span> <span class="fu">counts</span>(dds, <span class="at">normalized =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<blockquote class="blockquote">
<p>💡 <strong>NOTE</strong>: Although <code>DESeq2</code> does not use normalized counts as input (the normalization process happens inside automatically), the normalized counts we just generated are definitely useful when plotting results and performing clustering.</p>
</blockquote>
</section>
<section id="transforming-count-data" class="level2">
<h2 class="anchored" data-anchor-id="transforming-count-data">Transforming Count Data</h2>
<p>After we have normalized our data, we need to perform a <strong>transformation</strong>. There are many ways to transform count data but all of them achieve the goal of removing the gene-wise <strong>dependence between <em>variance</em> and <em>mean</em> of expression values</strong> across samples (something called <a href="https://en.wikipedia.org/wiki/Homoscedasticity_and_heteroscedasticity"><em>homoscedasticity</em></a>) in order to highlight interesting and biologically relevant expression trends even for genes expressed at lower values. We transform the data using a function provided in the <code>DESeq2</code> package called <code>rlog</code> which stands for <em>regularized logarithm</em>, without getting too much into the details of the workings of the function, we will <strong>transform the data</strong> and then look at how the gene-wise relationship between the <em>mean</em> and <em>variance</em> in our normalized data changes <strong>before and after</strong> the transformation.</p>
<p>The following code plot the mean/standard deviation relationship <strong>before the transformation</strong>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(vsn)</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot before data transformation</span></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a><span class="fu">meanSdPlot</span>(<span class="fu">assay</span>(dds))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="day2_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Transform the data and then plot the mean/standard deviation relationship <strong>after the transformation</strong>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Transform the data (watch how we create a new variable for it)</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>rld <span class="ot">&lt;-</span> <span class="fu">rlog</span>(dds)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check out the transformed values (notice how we now have floating point values and some are even negative!)</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">assay</span>(rld))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">

<table class="table" style="margin-left: auto; margin-right: auto;">
 <thead>
  <tr>
   <th style="text-align:left;">   </th>
   <th style="text-align:right;"> BSSE_QGF_204446 </th>
   <th style="text-align:right;"> BSSE_QGF_204447 </th>
   <th style="text-align:right;"> BSSE_QGF_204448 </th>
   <th style="text-align:right;"> BSSE_QGF_204449 </th>
   <th style="text-align:right;"> BSSE_QGF_204450 </th>
   <th style="text-align:right;"> BSSE_QGF_204451 </th>
   <th style="text-align:right;"> BSSE_QGF_204452 </th>
   <th style="text-align:right;"> BSSE_QGF_204453 </th>
   <th style="text-align:right;"> BSSE_QGF_204458 </th>
   <th style="text-align:right;"> BSSE_QGF_204459 </th>
   <th style="text-align:right;"> BSSE_QGF_204460 </th>
   <th style="text-align:right;"> BSSE_QGF_204461 </th>
   <th style="text-align:right;"> BSSE_QGF_204462 </th>
   <th style="text-align:right;"> BSSE_QGF_204463 </th>
   <th style="text-align:right;"> BSSE_QGF_204464 </th>
   <th style="text-align:right;"> BSSE_QGF_204465 </th>
  </tr>
 </thead>
<tbody>
  <tr>
   <td style="text-align:left;"> ENSG00000229851 </td>
   <td style="text-align:right;"> -0.3526715 </td>
   <td style="text-align:right;"> -0.2010792 </td>
   <td style="text-align:right;"> -0.3602817 </td>
   <td style="text-align:right;"> -0.3555970 </td>
   <td style="text-align:right;"> -0.2399632 </td>
   <td style="text-align:right;"> 0.0847179 </td>
   <td style="text-align:right;"> -0.3586727 </td>
   <td style="text-align:right;"> -0.2279047 </td>
   <td style="text-align:right;"> -0.3510900 </td>
   <td style="text-align:right;"> -0.0739337 </td>
   <td style="text-align:right;"> -0.1890122 </td>
   <td style="text-align:right;"> -0.3463143 </td>
   <td style="text-align:right;"> -0.2084209 </td>
   <td style="text-align:right;"> -0.3561915 </td>
   <td style="text-align:right;"> -0.0278781 </td>
   <td style="text-align:right;"> -0.2197011 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> ENSG00000228459 </td>
   <td style="text-align:right;"> 0.6365877 </td>
   <td style="text-align:right;"> 1.0374420 </td>
   <td style="text-align:right;"> 0.8520685 </td>
   <td style="text-align:right;"> 0.9875220 </td>
   <td style="text-align:right;"> 0.8502031 </td>
   <td style="text-align:right;"> 0.8730573 </td>
   <td style="text-align:right;"> 0.6243727 </td>
   <td style="text-align:right;"> 0.8723249 </td>
   <td style="text-align:right;"> 1.2186542 </td>
   <td style="text-align:right;"> 0.6423624 </td>
   <td style="text-align:right;"> 0.6506968 </td>
   <td style="text-align:right;"> 0.9392950 </td>
   <td style="text-align:right;"> 1.3012398 </td>
   <td style="text-align:right;"> 0.6294317 </td>
   <td style="text-align:right;"> 0.8565987 </td>
   <td style="text-align:right;"> 0.9948364 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> ENSG00000182888 </td>
   <td style="text-align:right;"> 4.2882919 </td>
   <td style="text-align:right;"> 4.3946526 </td>
   <td style="text-align:right;"> 4.6839023 </td>
   <td style="text-align:right;"> 3.8792719 </td>
   <td style="text-align:right;"> 4.2302617 </td>
   <td style="text-align:right;"> 3.9535142 </td>
   <td style="text-align:right;"> 4.3661274 </td>
   <td style="text-align:right;"> 4.0408208 </td>
   <td style="text-align:right;"> 4.7312951 </td>
   <td style="text-align:right;"> 4.2639791 </td>
   <td style="text-align:right;"> 4.3844344 </td>
   <td style="text-align:right;"> 3.9801799 </td>
   <td style="text-align:right;"> 4.3140752 </td>
   <td style="text-align:right;"> 4.2895333 </td>
   <td style="text-align:right;"> 4.2845088 </td>
   <td style="text-align:right;"> 4.4216270 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> ENSG00000286431 </td>
   <td style="text-align:right;"> 4.9014443 </td>
   <td style="text-align:right;"> 5.1148822 </td>
   <td style="text-align:right;"> 3.9461274 </td>
   <td style="text-align:right;"> 5.1565737 </td>
   <td style="text-align:right;"> 4.5110129 </td>
   <td style="text-align:right;"> 5.4814516 </td>
   <td style="text-align:right;"> 4.4043954 </td>
   <td style="text-align:right;"> 5.0560408 </td>
   <td style="text-align:right;"> 4.8115883 </td>
   <td style="text-align:right;"> 5.1016428 </td>
   <td style="text-align:right;"> 4.0314676 </td>
   <td style="text-align:right;"> 5.4787837 </td>
   <td style="text-align:right;"> 5.5748849 </td>
   <td style="text-align:right;"> 5.4050681 </td>
   <td style="text-align:right;"> 5.5932858 </td>
   <td style="text-align:right;"> 5.8469916 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> ENSG00000237490 </td>
   <td style="text-align:right;"> 0.3883848 </td>
   <td style="text-align:right;"> 0.5266944 </td>
   <td style="text-align:right;"> 0.3600419 </td>
   <td style="text-align:right;"> 0.2513714 </td>
   <td style="text-align:right;"> 0.3589418 </td>
   <td style="text-align:right;"> 0.4831765 </td>
   <td style="text-align:right;"> 0.4733968 </td>
   <td style="text-align:right;"> 0.3720291 </td>
   <td style="text-align:right;"> 0.6242683 </td>
   <td style="text-align:right;"> 0.6332632 </td>
   <td style="text-align:right;"> 0.6624091 </td>
   <td style="text-align:right;"> 0.4121075 </td>
   <td style="text-align:right;"> 0.3932301 </td>
   <td style="text-align:right;"> 0.2505595 </td>
   <td style="text-align:right;"> 0.3627161 </td>
   <td style="text-align:right;"> 0.2526425 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> ENSG00000101846 </td>
   <td style="text-align:right;"> 7.4157210 </td>
   <td style="text-align:right;"> 7.0617741 </td>
   <td style="text-align:right;"> 6.9775835 </td>
   <td style="text-align:right;"> 6.9710935 </td>
   <td style="text-align:right;"> 7.2251375 </td>
   <td style="text-align:right;"> 6.7612170 </td>
   <td style="text-align:right;"> 6.4012497 </td>
   <td style="text-align:right;"> 6.8805197 </td>
   <td style="text-align:right;"> 7.4628180 </td>
   <td style="text-align:right;"> 7.0709712 </td>
   <td style="text-align:right;"> 6.7358128 </td>
   <td style="text-align:right;"> 7.0828085 </td>
   <td style="text-align:right;"> 7.9416669 </td>
   <td style="text-align:right;"> 7.5904718 </td>
   <td style="text-align:right;"> 7.4141406 </td>
   <td style="text-align:right;"> 7.3517593 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> ENSG00000285679 </td>
   <td style="text-align:right;"> 3.1807531 </td>
   <td style="text-align:right;"> 3.0012482 </td>
   <td style="text-align:right;"> 3.7956542 </td>
   <td style="text-align:right;"> 3.9582976 </td>
   <td style="text-align:right;"> 2.5033603 </td>
   <td style="text-align:right;"> 3.1781599 </td>
   <td style="text-align:right;"> 3.3659313 </td>
   <td style="text-align:right;"> 3.3445709 </td>
   <td style="text-align:right;"> 2.8900663 </td>
   <td style="text-align:right;"> 3.2236691 </td>
   <td style="text-align:right;"> 3.0459007 </td>
   <td style="text-align:right;"> 3.2022524 </td>
   <td style="text-align:right;"> 3.4436592 </td>
   <td style="text-align:right;"> 2.9118579 </td>
   <td style="text-align:right;"> 3.7138348 </td>
   <td style="text-align:right;"> 3.4876301 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> ENSG00000219870 </td>
   <td style="text-align:right;"> -0.9180949 </td>
   <td style="text-align:right;"> -0.7722889 </td>
   <td style="text-align:right;"> -0.7014739 </td>
   <td style="text-align:right;"> -0.7917121 </td>
   <td style="text-align:right;"> -0.7029654 </td>
   <td style="text-align:right;"> -0.9207645 </td>
   <td style="text-align:right;"> -0.6933566 </td>
   <td style="text-align:right;"> -0.9208373 </td>
   <td style="text-align:right;"> -0.7779793 </td>
   <td style="text-align:right;"> -0.9137609 </td>
   <td style="text-align:right;"> -0.9060463 </td>
   <td style="text-align:right;"> -0.9071815 </td>
   <td style="text-align:right;"> -0.9167252 </td>
   <td style="text-align:right;"> -0.9202945 </td>
   <td style="text-align:right;"> -0.8037739 </td>
   <td style="text-align:right;"> -0.9193429 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> ENSG00000101849 </td>
   <td style="text-align:right;"> 11.7260434 </td>
   <td style="text-align:right;"> 11.9162085 </td>
   <td style="text-align:right;"> 11.3789329 </td>
   <td style="text-align:right;"> 11.6056330 </td>
   <td style="text-align:right;"> 11.7365594 </td>
   <td style="text-align:right;"> 12.0440052 </td>
   <td style="text-align:right;"> 11.7093709 </td>
   <td style="text-align:right;"> 11.8305500 </td>
   <td style="text-align:right;"> 11.4933369 </td>
   <td style="text-align:right;"> 11.7031860 </td>
   <td style="text-align:right;"> 11.5560969 </td>
   <td style="text-align:right;"> 11.5576538 </td>
   <td style="text-align:right;"> 11.6527863 </td>
   <td style="text-align:right;"> 11.9395408 </td>
   <td style="text-align:right;"> 11.6311480 </td>
   <td style="text-align:right;"> 11.7118694 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> ENSG00000047644 </td>
   <td style="text-align:right;"> 12.1678896 </td>
   <td style="text-align:right;"> 12.2710704 </td>
   <td style="text-align:right;"> 11.7648111 </td>
   <td style="text-align:right;"> 12.2494452 </td>
   <td style="text-align:right;"> 12.2612293 </td>
   <td style="text-align:right;"> 12.2261545 </td>
   <td style="text-align:right;"> 12.1440675 </td>
   <td style="text-align:right;"> 12.3066751 </td>
   <td style="text-align:right;"> 12.2295272 </td>
   <td style="text-align:right;"> 12.2089019 </td>
   <td style="text-align:right;"> 12.1077662 </td>
   <td style="text-align:right;"> 12.3206846 </td>
   <td style="text-align:right;"> 12.2239945 </td>
   <td style="text-align:right;"> 12.3466937 </td>
   <td style="text-align:right;"> 11.8984296 </td>
   <td style="text-align:right;"> 12.3830057 </td>
  </tr>
</tbody>
</table>

</div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># let's plot the transformed values</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="fu">meanSdPlot</span>(<span class="fu">assay</span>(rld))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="day2_files/figure-html/unnamed-chunk-18-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<blockquote class="blockquote">
<p>💡 <strong>It is clear how now genes with high mean expression values (on the right) are comparable in terms of standard deviation to genes with lower mean expression values (on the left).</strong></p>
</blockquote>
</section>
<section id="data-quality-visual-assessment" class="level2">
<h2 class="anchored" data-anchor-id="data-quality-visual-assessment">Data Quality Visual Assessment</h2>
<p>One way to understand trends in our data and the present of poor quality or <strong><em>outlier</em></strong> samples is to perform exploratory analyses through visualization. In <code>R</code> in general, data visualization is aided by the presence of many packages (on top the basic plotting functionality) which can handle diverse kinds of data visualization tasks (from traditional plots to <strong>visualizing tabular data through heatmaps</strong>). We will encounter two of these packages, one is <a href="https://ggplot2.tidyverse.org/"><code>ggplot2</code></a> and the other one is <a href="https://cran.r-project.org/web/packages/pheatmap/index.html"><code>pheatmap</code></a>.</p>
<section id="clustering" class="level3">
<h3 class="anchored" data-anchor-id="clustering">Clustering</h3>
<p>One of the main strategies for checking the consistency of our dataset is to <strong>cluster samples based on their complete expression profile</strong> (which as you might recall consists of 31737 genes in our dataset). This will allow us to <strong>spot the presence of <em>outliers</em> in the data</strong> and understand if we see consistent profiles of gene expression across replicates, which we expect. Use the code below to plot a <em>heatmap</em> of normalized expression values for our samples. Since plotting the full expression table can be computationally expensive, we might want to subset it to the 200 top expressed genes in the dataset.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"pheatmap"</span>)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Take the top 200 genes in the dataset</span></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>select <span class="ot">&lt;-</span> <span class="fu">order</span>(<span class="fu">rowMeans</span>(<span class="fu">counts</span>(dds,<span class="at">normalized=</span><span class="cn">TRUE</span>)),</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>                <span class="at">decreasing=</span><span class="cn">TRUE</span>)[<span class="dv">1</span><span class="sc">:</span><span class="dv">200</span>] <span class="co"># Select number of genes</span></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Create another table for annotating the heatmap with colors</span></span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">colData</span>(dds)[,<span class="fu">c</span>(<span class="st">"Donor"</span>,<span class="st">"SampleGroup"</span>)])</span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Draw the heatmap using the `pheatmap` package</span></span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(<span class="fu">assay</span>(rld)[select,], <span class="at">cluster_rows=</span><span class="cn">FALSE</span>, <span class="at">show_rownames=</span><span class="cn">FALSE</span>,</span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a>         <span class="at">cluster_cols=</span><span class="cn">TRUE</span>, <span class="at">annotation_col=</span>df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="day2_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<blockquote class="blockquote">
<p>💡 <strong>What type of assessment would you make about the consistency of the samples across these top 200 genes? Do they cluster (a synonym for <em>similar</em>) based on the donor or on the biological condition of our interest?</strong></p>
</blockquote>
</section>
<section id="sample-to-sample-distances" class="level3">
<h3 class="anchored" data-anchor-id="sample-to-sample-distances">Sample-to-sample Distances</h3>
<p>Another way to <strong>get a sense of the global relationship between samples</strong> is to check for how <em>distant</em> samples are between themselves. This analysis of <em>pairwise</em> distances looks at the expression value of all 31737 genes in the dataset and determines which samples have a more or less similar or different expression value for each. <strong>We expect biologically similar samples to have very little difference</strong>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(RColorBrewer)</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute distances</span></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>sampleDists <span class="ot">&lt;-</span> <span class="fu">dist</span>(<span class="fu">t</span>(<span class="fu">assay</span>(rld)))</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Organize</span></span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>sampleDistMatrix <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(sampleDists)</span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a>colors <span class="ot">&lt;-</span> <span class="fu">colorRampPalette</span>( <span class="fu">rev</span>(<span class="fu">brewer.pal</span>(<span class="dv">9</span>, <span class="st">"Blues"</span>)) )(<span class="dv">255</span>)</span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot with `pheatmap`</span></span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(sampleDistMatrix,</span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a>         <span class="at">clustering_distance_rows=</span>sampleDists,</span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true" tabindex="-1"></a>         <span class="at">clustering_distance_cols=</span>sampleDists,</span>
<span id="cb19-15"><a href="#cb19-15" aria-hidden="true" tabindex="-1"></a>         <span class="at">color =</span> colors,</span>
<span id="cb19-16"><a href="#cb19-16" aria-hidden="true" tabindex="-1"></a>         <span class="at">annotation_col =</span> df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="day2_files/figure-html/unnamed-chunk-20-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<blockquote class="blockquote">
<p>💡 <strong>What type of assessment would you make about the heatmap you just produced? Which CD8 T-cell populations are mostly similar in terms of overall gene expression profile? Does a particular population stand out?</strong></p>
</blockquote>
</section>
<section id="principal-component-analysis-pca" class="level3">
<h3 class="anchored" data-anchor-id="principal-component-analysis-pca">Principal Component Analysis (PCA)</h3>
<p>Another useful approach for <strong>understanding the main variability axes in our data</strong> is to compute and plot a <strong><a href="https://en.wikipedia.org/wiki/Principal_component_analysis">PCA</a></strong>. Without getting into the details, PCA takes our expression data and outputs its <strong>principal components</strong>, which encode the main sources of variability in the data. Ideally, <strong>we want the samples have variability caused by the biological effect of our interest</strong> (in this case the differences between CD8 T-cell populations), but this might not be the case. By plotting and coloring the points by different <strong><em>covariates</em></strong> (i.e.&nbsp;donor or cell type) we are able to understand where the variability comes from and if there is any detectable <strong><a href="https://en.wikipedia.org/wiki/Batch_effect">batch effect</a></strong>. Use the code below to generate two <em>scatter plots</em> of PCA coordinates and color them to understand what causes the variability in the data.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate principal components and percentage of variance</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>pcaData <span class="ot">&lt;-</span> <span class="fu">plotPCA</span>(rld, <span class="at">intgroup=</span><span class="fu">c</span>(<span class="st">"Donor"</span>,<span class="st">"SampleGroup"</span>), <span class="at">returnData=</span><span class="cn">TRUE</span>)</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>percentVar <span class="ot">&lt;-</span> <span class="fu">round</span>(<span class="dv">100</span> <span class="sc">*</span> <span class="fu">attr</span>(pcaData, <span class="st">"percentVar"</span>))</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot (this time with ggplot2!!)</span></span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(pcaData, <span class="fu">aes</span>(PC1, PC2, <span class="at">color=</span>SampleGroup, <span class="at">shape=</span>Donor)) <span class="sc">+</span></span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size=</span><span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="fu">paste0</span>(<span class="st">"PC1: "</span>,percentVar[<span class="dv">1</span>],<span class="st">"% variance"</span>)) <span class="sc">+</span></span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="fu">paste0</span>(<span class="st">"PC2: "</span>,percentVar[<span class="dv">2</span>],<span class="st">"% variance"</span>)) <span class="sc">+</span> </span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_fixed</span>() <span class="sc">+</span></span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="day2_files/figure-html/unnamed-chunk-21-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<blockquote class="blockquote">
<p>💡 <strong>What can you say about this PCA? Are the samples positioning themselves based on the donor or their biological condition? What is the most extreme group of samples? How does this information relate to the other plots we produced above?</strong></p>
</blockquote>
<p>Now that we have plotted all the main diagnostic information related to the dataset and we have a bit of a grasp of it, we can start thinking about <strong>testing for differentially expressed genes</strong>.</p>


</section>
</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
      <div class="nav-footer-center">Copyright 2023, Jacopo Arrigoni &amp; Mattia Toninelli</div>
  </div>
</footer>



</body></html>